### ISSUE TITLE
Docs blocks

### ISSUE THREAD

**Issue by drewbanin** [fishtown-analytics/dbt#810]
### Overview

In support of autogenerated dbt project documentation (https://github.com/fishtown-analytics/dbt/issues/375, https://github.com/fishtown-analytics/dbt/issues/790), dbt needs a construct to represent long-form, markdown documentation.

This construct should take the form of a Jinja block, `{% docs %}`. These blocks will contain text that will be interpreted as Markdown, and should be compiled with a sensible context.

### Examples

These docs blocks should be named so that they can be referenced from elsewhere in the dbt project. For example, a model can be documented by building the following docs block:

**models/snowplow/sessions/readme.md**
```
{% docs snowplow_sessions %}

This table contains one record for every session recorded by Snowplow.
A session is itself comprised of pageviews that all occur within 30 minutes
of each other. If more than 30 minutes elapse between pageviews, then a
new session is created. Given the following pageviews:

| session_id | page_view_id | page_title |
| ---------- | ------------ | ---------- |
| abc        | 123          | Home       |
| abc        | 456          | About      |
| abc        | 789          | Home       |

The following sessions will be created:

| session_id | first_page_title | count_pageviews |
| ---------- | ---------------- | --------------- |
| abc        | 123              | 2               |
| abc        | 789              | 1               |

{% enddocs %}

{% docs snowplow_sessions__session_id %}

This column is the unique identifier for a Snowplow session. It is generated by a cookie
then expires after 30 minutes of inactivity.

{% enddocs %}
```

Because these blocks are named, they can be referenced elsewhere in the dbt project, for example, in a `schema.yml` file (https://github.com/fishtown-analytics/dbt/issues/790). Eg:

```yml
snowplow_sessions:
  comment: {{ ref('snowplow_sessions', type='docs') }}
  columns:
    - name: session_id
      comment: {{ ref('snowplow_sessions__session_id', type='docs') }}
```

The linkage between a model schema and documentation can be handled in a couple of ways:
1. Purely on a name basis, as shown above
2. With a `model=` config in the docs block

An example of approach (2) might look like:

```
{% docs model="snowplow_sessions" %}

... docs ...

{% enddocs %}
```

If this approach is used, then it may not be necessary to specify the comment for the model in `schema.yml`. I sort of dislike this approach, as the docs could become fragmented across many different files. I'm definitely open to discussion though.

One other benefit of defining documentation separately from a model definition is that docs can be reused where appropriate. The `session_id` column is used in _many_ places, both inside and outside of the Snowplow package. I _really_ like the idea of being able to use *exactly* the same definition for this column everywhere that it appears. For example:

```yml
# A model in the Snowplow package
snowplow_sessions:
  columns:
    - name: session_id
      comment: {{ ref('snowplow', 'session_id', type='docs') }}

....

# A model in my project
session_attribution:
  columns:
    - name: session_id
      comment: {{ ref('snowplow', 'session_id', type='docs') }}
...
```

### Referencing docs

Above, I used the syntax `{{ ref(package_name, model_name, type) }}`. The two-argument version of `ref` currently exists (package name + model name), but the `type` argument would be new. I think we have a few options:
1. Use `ref`, but add an argument to specify a `type`, as shown above
2. Use `ref`, but infer what types of resources can be referenced by the context of its use
3. Use a different function like `docs(....)` which wraps some underlying implementation of `ref`, but only selects docs nodes. We could call this `ref_docs` if that seems more consistent?

I think that (3) above is my preference, but I am open to discussion.

### Compilation Context

Some dbt concepts map naturally to the docs environment. For example, there's a great opportunity to use macros in docs to eg. reuse a documentation template, generate a markdown table, add links, etc. These macros could conceivably live in a dbt package. Cool!

Other concepts like `ref()` are interesting to consider. I can imagine that if you `{{ ref(...) }}` a model inside of the docs, dbt could compile in `schema.table`, so that the docs would be correct regardless of compilation environment. I can also imagine making it possible to generate _links_ to other dbt models when `ref()` is used. The context available to users during compilation should be sensible and consistent with the model context, to the extent possible. 

cc @jthandy @cmcarthur

**Comment by jthandy** [comment]
1. I agree--I do not think you should be able to specify what object is being documented within a docs block. I believe this because a) I think the object itself should specifically call docs (more readable), and b) I think it would be great to have docs reference-able in multiple locations as you indicate. I think of a docs block as essentially just a markdown blob with a unique identifier.

2. Your option #3 is my preferred way of referencing docs as well. I would vote for just `docs(package_name, docs_name)`. I know that, under the hood, everything is just nodes and `ref()` and `docs()` are very similar functions, but I believe that the user intention is quite different when calling these two functions and I think we would do well to indicate to the user that they are different. I would prefer not to have `ref` be a part of the method name for this at all—from a user's perspective, ref is what you do when you are pointing to a model.

3. I like what you're saying re: context. An additional point I'd like to raise here is `vars`. I think that it could be quite important to document variable values--especially constants that get passed in. I frequently pass in constants to models, and it would be great if I could reference those in the docs markdown as well. This might be the one reason that it would be important to start defining variables within the model itself instead of just in `dbt_project.yml`--you might want to define local constants in the model so that they can be referenced by docs as well.

**Comment by cmcarthur** [comment]
> 2. With a model= config in the docs block
>
> An example of approach (2) might look like:
>
> ```
> {% docs model="snowplow_sessions" %}
>
> ... docs ...
>
> {% enddocs %}
> ```
> If this approach is used, then it may not be necessary to specify the comment for the model in schema.yml. I sort of dislike this approach, as the docs could become fragmented across many different files. I'm definitely open to discussion though.

I agree that this approach is poor -- it combines the bad qualities of both options. The docs aren't right next to the model that is being documented, and you can't reuse this documentation elsewhere.

A viable alternative to this would be inline documentation:

```model.sql
{{config(materialized='view')}}

{% docs %}
... docs for this model ...
{% enddocs %}

select * from ...
```

This can't be reused, but it deals with the fragmentation issue. I don't think we should attempt this right now. IMO the only way to link docs to models for the immediate future should be:

```
...
  comment: {{ doc('model_name') }}
...
```

**Comment by cmcarthur** [comment]
Actually come to think of it I think the inline method above would be a nice way to add docs to constructs that aren't in schema.yml (macros? vars?)

**Comment by cmcarthur** [comment]
@drewbanin I agree with all of the preferred methods you described above:

- use the `comment:` section of schema.yml to link doc blocks to models
- use a new function to reference doc blocks, rather than `ref` with an additional argument

it sounds like we should just take this approach.

to your point about compilation context, things like calling macros in schema.yml files should just work. this relates to what we were discussing yesterday regarding parsing & post processing. i'm writing something else up about this, i'll post it elsewhere.

finally, i feel a little bit of discomfort with leaning on jinja more and more heavily, since i don't think it's a particularly good way to organize code. i can't think of a better way given the tools we have at our fingertips.... but i just want to acknowledge that we're making an active decision to lean on jinja more

**Comment by beckjake** [comment]
Closed by #888
